ddtrace/contrib/langchain/patch.py

```diff
@@ -0,0 +1,752 @@
+from ddtrace.contrib.trace_utils import with_traced_module
+from ddtrace.contrib.trace_utils import wrap
+from ddtrace.internal.agent import get_stats_url
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.internal.utils.formats import asbool
+from ddtrace.pin import Pin
+
+
+if TYPE_CHECKING:
+    from ddtrace import Span
+
+
+log = get_logger(__name__)
+
+
+config._add(
+    "langchain",
+    {
+        "logs_enabled": asbool(os.getenv("DD_LANGCHAIN_LOGS_ENABLED", False)),
+        "metrics_enabled": asbool(os.getenv("DD_LANGCHAIN_METRICS_ENABLED", True)),
+        "span_prompt_completion_sample_rate": float(os.getenv("DD_LANGCHAIN_SPAN_PROMPT_COMPLETION_SAMPLE_RATE", 1.0)),
+        "log_prompt_completion_sample_rate": float(os.getenv("DD_LANGCHAIN_LOG_PROMPT_COMPLETION_SAMPLE_RATE", 0.1)),
+        "span_char_limit": int(os.getenv("DD_LANGCHAIN_SPAN_CHAR_LIMIT", 128)),
+        "_api_key": os.getenv("DD_API_KEY"),
+    },
+)
+
+
+class _LangChainIntegration(BaseLLMIntegration):
+    _integration_name = "langchain"
+
+    def __init__(self, config, stats_url, site, api_key):
+        super().__init__(config, stats_url, site, api_key)
+
+    def _set_base_span_tags(self, span, interface_type="", provider=None, model=None, api_key=None):
+        # type: (Span, str, Optional[str], Optional[str], Optional[str]) -> None
+        """Set base level tags that should be present on all LangChain spans (if they are not None)."""
+        span.set_tag_str(TYPE, interface_type)
+        if provider is not None:
+            span.set_tag_str(PROVIDER, provider)
+        if model is not None:
+            span.set_tag_str(MODEL, model)
+        if api_key is not None:
+            if len(api_key) >= 4:
+                span.set_tag_str(API_KEY, "...%s" % str(api_key[-4:]))
+            else:
+                span.set_tag_str(API_KEY, api_key)
+
+    @classmethod
+    def _logs_tags(cls, span):
+        # type: (Span) -> str
+        api_key = span.get_tag("langchain.request.api_key") or ""
+        tags = "env:%s,version:%s,langchain.request.provider:%s,langchain.request.model:%s,langchain.request.type:%s,langchain.request.api_key:%s" % (  # noqa: E501
+            (config.env or ""),
+            (config.version or ""),
+            (span.get_tag(PROVIDER) or ""),
+            (span.get_tag("langchain.request.model") or ""),
+            (span.get_tag("langchain.request.type") or ""),
+            api_key,
+        )
+        return tags
+
+    @classmethod
+    def _metrics_tags(cls, span):
+        # type: (Span) -> list
+        provider = span.get_tag(PROVIDER) or ""
+        api_key = span.get_tag("langchain.request.api_key") or ""
+        tags = [
+            "version:%s" % (config.version or ""),
+            "env:%s" % (config.env or ""),
+            "service:%s" % (span.service or ""),
+            "langchain.request.provider:%s" % provider,
```

Comment:
nit: We can format some of these strings to use the provider, type and model constants

Ex:
```suggestion
            "%s:%s" % (PROVIDER, provider),
```